### PR TITLE
feat: Hyperlane AVS contracts

### PR DESCRIPTION
**Pull request by aroralanuk**

### CODE REVIEW

File: solidity/remappings.txt [hunk 1]
```diff
@@ -1,5 +1,8 @@
 @openzeppelin=../node_modules/@openzeppelin
+@openzeppelin/contracts-upgradeable/=../node_modules/@openzeppelin/contracts-upgradeable/
```
Comment (yorhodes) [inline solidity/remappings.txt:2]: doesnt the above line already resolve this?

Comment (aroralanuk) [reply solidity/remappings.txt:2]: Unfortunately not. If I don't explicitly remap it, then it uses the upgradable contracts from inside the lib/eigenlayer-core sub module. Check the below line.

**Review by yorhodes — CHANGES_REQUESTED** [review]
didnt look at tests yet
please fix the build, obviously that will block deployment

File: solidity/contracts/libs/EnumerableMapEnrollment.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/libs/EnumerableMapEnrollment.sol:1]: Yes, that's correct.

File: solidity/contracts/libs/EnumerableMapEnrollment.sol [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    enrollment.unenrollmentStartBlock
+                )
+            );
+    }
+
+    function decode(bytes32 encoded) public view returns (Enrollment memory) {
+        uint8 status = uint8(encoded[0]);
+        uint248 unenrollmentStartBlock = uint248(uint256((encoded << 8) >> 8));
+        return Enrollment(EnrollmentStatus(status), unenrollmentStartBlock);
+    }
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/libs/EnumerableMapEnrollment.sol:51]: not sure if I understand, I don't know if we can extend a EnumerableMap with a struct like EnumerableMap.Bytes32ToBytes _inner;

File: solidity/contracts/Mailbox.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/Mailbox.sol:1]: reverted

File: solidity/contracts/avs/ECDSAServiceManagerBase.sol [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import {IServiceManager} from "@eigenlayer/middleware/interfaces/IServiceManager.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+contract ECDSAServiceManagerBase is IServiceManager, OwnableUpgradeable {
+    // ============ Constants ============
+
+    // Stake registry responsible for maintaining operator stakes
+    ECDSAStakeRegistry internal immutable stakeRegistry;
+    // Eigenlayer's AVS directory for interactions between AVS and operators
+    IAVSDirectory internal immutable elAvsDirectory;
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/ECDSAServiceManagerBase.sol:32]: avsDirectory is a function in IServiceManager which returns an address so it won't compile bc of the conflicting return type.  Either it's the prefix or I can change it into `address internal immutable elAvsDirectory;` which ever you prefer.

File: solidity/contracts/avs/ECDSAServiceManagerBase.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/ECDSAServiceManagerBase.sol:1]: Yes.

File: solidity/contracts/test/ERC4626/ERC4626Test.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/test/ERC4626/ERC4626Test.sol:1]: I get "Invalid type for argument in modifier invocation. Invalid implicit conversion from contract IERC20 to contract IERC20Metadata requested." error.  It's a remapping issue where el has a older OZ version but I'm mistakely using for our contracts.

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+     */
+    function queueUnenrollmentFromChallengers(
+        IRemoteChallenger[] memory _challengers
+    ) external {
+        for (uint256 i = 0; i < _challengers.length; i++) {
+            IRemoteChallenger challenger = _challengers[i];
+            (bool exists, Enrollment memory enrollment) = enrolledChallengers[
+                msg.sender
+            ].tryGet(address(challenger));
+            if (exists && enrollment.status == EnrollmentStatus.ENROLLED) {
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:128]: fixed

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            }
+        }
+    }
+
+    /**
+     * @notice Completes the unenrollment of an operator from a list of challengers
+     * @param operator The address of the operator
+     * @param _challengers The list of challengers to unenroll from
+     */
+    function completeQueuedUnenrollmentFromChallengers(
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:153]: fixed

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+                    address(challenger),
+                    Enrollment(EnrollmentStatus.ENROLLED, 0)
+                )
+            );
+            emit OperatorEnrolledToChallenger(msg.sender, challenger);
+        }
+    }
+
+    /**
+     * @notice Queues an operator for unenrollment from a list of challengers
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:117]: fixed

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    ) public onlyStakeRegistryOrOperator(operator) {
+        for (uint256 i = 0; i < _challengers.length; i++) {
+            IRemoteChallenger challenger = _challengers[i];
+            (bool exists, Enrollment memory enrollment) = enrolledChallengers[
+                operator
+            ].tryGet(address(challenger));
+
+            require(
+                exists &&
+                    enrollment.status != EnrollmentStatus.ENROLLED &&
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:165]: fixed

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+                    block.number >=
+                    enrollment.unenrollmentStartBlock +
+                        challenger.challengeDelayBlocks(),
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:168]: Good question, I think either have it as immutable in the challenger or if it's changeable you have to trust the challenger deployer.
Maybe we have a max time period for this in the HSM.

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        address[] memory keys = enrolledChallengers[_operator].keys();
+
+        IRemoteChallenger[] memory challengers = new IRemoteChallenger[](
+            keys.length
+        );
+        for (uint256 i = 0; i < keys.length; i++) {
+            challengers[i] = IRemoteChallenger(keys[i]);
+        }
+        return challengers;
+    }
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:212]: they I need to convert from address[] memory to IRemoteChallenger[] memory for freezeOperator

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+    // Mapping of operators to challengers they are enrolled in (enumerable required for remove-all)
+    mapping(address => EnumerableMapEnrollment.AddressToEnrollmentMap)
+        internal enrolledChallengers;
+
+    // ============ Modifiers ============
+
+    // Only allows the challenger the operator is enrolled in to call the function
+    modifier onlyEnrolledChallenger(address operator) {
+        (bool exists, ) = enrolledChallengers[operator].tryGet(msg.sender);
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:79]: that's why I made the enum:

```solidity
enum EnrollmentStatus {
    UNENROLLED,
    ENROLLED,
    PENDING_UNENROLLMENT
}
```

so UNENROLLED is falsely without me explicitly setting it anywhere

File: solidity/remappings.txt [hunk 11]
```diff
@@ -1,5 +1,11 @@
...
+@openzeppelin/contracts/=lib/eigenlayer-middleware/lib/openzeppelin-contracts/contracts/
 @layerzerolabs=../node_modules/@layerzerolabs
 @eth-optimism=../node_modules/@eth-optimism
 ds-test/=lib/forge-std/lib/ds-test/src/
 forge-std/=lib/forge-std/src/
+@eigenlayer/core/=lib/eigenlayer-contracts/src/contracts/core/
+@eigenlayer/ecdsa/=lib/eigenlayer-middleware/src/unaudited/
+@eigenlayer/middleware/=lib/eigenlayer-middleware/src/
+@eigenlayer/interfaces/=lib/eigenlayer-contracts/src/contracts/interfaces/
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/remappings.txt:11]: fixed

File: solidity/contracts/test/TestRemoteChallenger.sol [hunk 12]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+import {IRemoteChallenger} from "../interfaces/avs/IRemoteChallenger.sol";
+import {HyperlaneServiceManager} from "../avs/HyperlaneServiceManager.sol";
+
+contract TestRemoteChallenger is IRemoteChallenger {
+    HyperlaneServiceManager internal immutable hsm;
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/test/TestRemoteChallenger.sol:8]: Yes, the challenger shouldn't be able to change the associated HSM

File: solidity/contracts/libs/EnumerableMapEnrollment.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/libs/EnumerableMapEnrollment.sol:53]: oh I see
I guess EnumerableMap is a mapping to a Set and we cant have a Set with this struct as the value type?

File: solidity/contracts/avs/HyperlaneServiceManager.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:79]: gotcha
so this is implicity converting between your enum and bool? that seems dangerous

[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:79]: isnt it possible for `enrollment == UNENROLLED` and `exists == true`?

File: solidity/contracts/avs/ECDSAServiceManagerBase.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/ECDSAServiceManagerBase.sol:1]: done

File: solidity/contracts/avs/HyperlaneServiceManager.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:85]: no, when status == UNENROLLED, we reset unenrollmentStartBlock

File: solidity/script/DeployAVS.s.sol [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+            string(abi.encodePacked(".", targetEnv, ".avsDirectory"))
+        );
+        delegationManager = IDelegationManager(
+            json.readAddress(
+                string(abi.encodePacked(".", targetEnv, ".delegationManager"))
+            )
+        );
+        paymentCoordinator = json.readAddress(
+            string(abi.encodePacked(".", targetEnv, ".paymentCoordinator"))
+        );
```
Comment (yorhodes) [inline solidity/script/DeployAVS.s.sol:37]: why not `vm.parseJson` and abi.decode

File: solidity/script/DeployAVS.s.sol [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    function run() external {
+        vm.startBroadcast();
+
+        loadEigenlayerAddresses();
```
Comment (yorhodes) [inline solidity/script/DeployAVS.s.sol:43]: nit: load addresses before starting broadcast?

File: solidity/script/DeployAVS.s.sol [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+        ECDSAStakeRegistry stakeRegistry = new ECDSAStakeRegistry(
+            delegationManager
+        );
```
Comment (yorhodes) [inline solidity/script/DeployAVS.s.sol:47]: is this not already deployed?

File: solidity/script/DeployAVS.s.sol [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        HyperlaneServiceManager hsm = new HyperlaneServiceManager(
+            avsDirectory,
+            address(stakeRegistry),
+            paymentCoordinator,
+            address(delegationManager)
+        );
+
+        stakeRegistry.initialize(address(hsm), thresholdWeight, quorum);
```
Comment (yorhodes) [inline solidity/script/DeployAVS.s.sol:55]: add a proxy admin address to config and deploy this with a proxy for upgradeability?

File: solidity/test/avs/HyperlaneServiceManager.t.sol [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+                delegationApprover: address(0),
+                stakerOptOutWindowBlocks: 0
+            }),
+            ""
+        );
+        // set operator as registered in Eigenlayer
+        delegationManager.setIsOperator(operator, true);
+    }
+
+    event AVSMetadataURIUpdated(address indexed avs, string metadataURI);
```
Comment (yorhodes) [inline solidity/test/avs/HyperlaneServiceManager.t.sol:74]: I believe you can share events across files now
https://github.com/ethereum/solidity/releases/tag/v0.8.22

File: solidity/test/avs/HyperlaneServiceManager.t.sol [hunk 18]
```diff
@@ -0,0 +1,470 @@
...
+        // assert
+        IAVSDirectory.OperatorAVSRegistrationStatus operatorStatus = avsDirectory
+                .avsOperatorStatus(address(_hsm), operator);
+        assertEq(
+            uint8(operatorStatus),
+            uint8(IAVSDirectory.OperatorAVSRegistrationStatus.UNREGISTERED)
+        );
+    }
+
+    /// forge-config: default.fuzz.runs = 10
```
Comment (yorhodes) [inline solidity/test/avs/HyperlaneServiceManager.t.sol:155]: cool!

File: solidity/test/avs/HyperlaneServiceManager.t.sol [hunk 19]
```diff
@@ -0,0 +1,470 @@
...
+        for (uint256 i = 0; i < challengers.length; i++) {
+            vm.expectCall(
+                address(slasher),
+                abi.encodeCall(ISlasher.freezeOperator, (operator))
+            );
+            challengers[i].handleChallenge(operator);
+        }
+    }
+
+    /// forge-config: default.fuzz.runs = 10
```
Comment (yorhodes) [inline solidity/test/avs/HyperlaneServiceManager.t.sol:305]: should we just configure this in the foundry.toml? whats the default number of fuzz runs?

**Review by yorhodes — APPROVED** [review]
lgtm

**Review by yorhodes — CHANGES_REQUESTED** [review]
sorry will need one more round of review and a bit closer look at your tests before approval

File: solidity/contracts/libs/EnumerableMapEnrollment.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/libs/EnumerableMapEnrollment.sol:1]: I didn't wanna clutter the contract itself.

File: solidity/contracts/libs/EnumerableMapEnrollment.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/libs/EnumerableMapEnrollment.sol]: No, it's possible but we won't be able to use any of OZ libraries and will need to write it from scratch.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/52c36d412e8681053975396223d0ea39687fe33b/contracts/utils/structs/EnumerableSet.sol#L41

File: solidity/contracts/avs/ECDSAServiceManagerBase.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/ECDSAServiceManagerBase.sol]: actually, steven changed it to an address and that fixes it.

File: solidity/contracts/avs/HyperlaneServiceManager.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:82]: No, it does. The value is 0x0000..0000 which is false. I can add a test which sets it as UNENROLLED and then, check if exists == false.

File: solidity/contracts/avs/ECDSAStakeRegistry.sol [hunk 20]
```diff
@@ -0,0 +1,535 @@
...
+    /// @return The function selector that indicates the signature is valid according to ERC1271 standard.
+    function isValidSignature(
+        bytes32 _dataHash,
+        bytes memory _signatureData
+    ) external view returns (bytes4) {
+        (
+            address[] memory signers,
+            bytes[] memory signatures,
+            uint32 referenceBlock
+        ) = abi.decode(_signatureData, (address[], bytes[], uint32));
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/ECDSAStakeRegistry.sol:119]: This is not my code, copied from https://github.com/Layr-Labs/eigenlayer-middleware/blob/dev/src/unaudited/ECDSAServiceManagerBase.sol

File: solidity/contracts/test/ERC4626/ERC4626Test.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/test/ERC4626/ERC4626Test.sol:1]: fixed

File: solidity/contracts/avs/HyperlaneServiceManager.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol]: ok, I switched to using address[] map in completeUnenrollment. Let me know if I start use address[] in enroll and startUnenrollment functions too.

[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol]: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/828f4e9846909f33cd6e5fbe4b40205c33790ded

File: solidity/contracts/avs/HyperlaneServiceManager.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:82]: added test here: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/b50e0b25a1ac181ac8be8d86b0eec76b4fe95aca

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+        __ServiceManagerBase_init(msg.sender);
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Enrolls as an operator into a list of challengers
+     * @param _challengers The list of challengers to enroll into
+     */
+    function enrollIntoChallengers(
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:115]: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/7112431989b06ed9fc6fcf051f825601759e3853

File: solidity/contracts/avs/HyperlaneServiceManager.sol [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+            );
+            emit OperatorEnrolledToChallenger(msg.sender, challenger);
+        }
+    }
+
+    /**
+     * @notice starts an operator for unenrollment from a list of challengers
+     * @param _challengers The list of challengers to unenroll from
+     */
+    function startUnenrollment(
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/avs/HyperlaneServiceManager.sol:127]: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/7112431989b06ed9fc6fcf051f825601759e3853

File: solidity/contracts/interfaces/avs/IRemoteChallenger.sol
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/contracts/interfaces/avs/IRemoteChallenger.sol:1]: separated into a vendor folder: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/4cf5b0203e1be3a9225034f12e4d5b35a97bf5fb

File: solidity/script/DeployAVS.s.sol [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+    address public paymentCoordinator;
+
+    Quorum quorum;
+    uint256 thresholdWeight = 6667;
+
+    function loadEigenlayerAddresses(string targetEnv) external {
+        string memory root = vm.projectRoot();
+        string memory path = string.concat(
+            root,
+            "solidity/script/avs/eigenlayer_addresses.json"
```
[the remark this answers is not in the document]
Comment (aroralanuk) [reply solidity/script/DeployAVS.s.sol]: where should I put it?

File: solidity/script/DeployAVS.s.sol (hunk 13 above)
Comment (aroralanuk) [reply solidity/script/DeployAVS.s.sol]: readAddress does parseJson behind the scenes, it's a nice helper. I'm using parseJson for reading the StrategyInfo.

File: solidity/script/DeployAVS.s.sol (hunk 14 above)
Comment (aroralanuk) [reply solidity/script/DeployAVS.s.sol]: fixed

File: solidity/script/DeployAVS.s.sol (hunk 15 above)
Comment (aroralanuk) [reply solidity/script/DeployAVS.s.sol]: stakeregistry is separate for every AVS

File: solidity/test/avs/HyperlaneServiceManager.t.sol (hunk 17 above)
Comment (aroralanuk) [reply solidity/test/avs/HyperlaneServiceManager.t.sol:75]: Ahh nice. But, can we rely on >=0.8.22 given our contracts can also work for >=0.8.0?

File: solidity/script/DeployAVS.s.sol (hunk 16 above)
Comment (aroralanuk) [reply solidity/script/DeployAVS.s.sol]: fixed here: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/7f3540bfa1feb031336695c9f01a0c812609bd7f

File: solidity/test/avs/HyperlaneServiceManager.t.sol (hunk 19 above)
Comment (aroralanuk) [reply solidity/test/avs/HyperlaneServiceManager.t.sol]: 256 I believe but for local testing it was too slow. I can remove it.

Comment (aroralanuk) [reply solidity/test/avs/HyperlaneServiceManager.t.sol]: switched to global fuzz.runs = 50 and higher dictionary weight which means it will looks for edge case values more frequently here https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3651/commits/23aa15529a22408b9baef54bff151b484f249937

File: solidity/script/avs/DeployAVS.s.sol [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+                string(abi.encodePacked(".", targetEnv, ".avsDirectory"))
+            )
+        );
+        delegationManager = IDelegationManager(
+            json.readAddress(
+                string(abi.encodePacked(".", targetEnv, ".delegationManager"))
+            )
+        );
+        // paymentCoordinator = IPaymentCoordinator(json.readAddress(string(abi.encodePacked(".", targetEnv, ".paymentCoordinator"))));
+        paymentCoordinator = new TestPaymentCoordinator(); // temporary until Eigenlayer deploys the real one
```
Comment (yorhodes) [inline solidity/script/avs/DeployAVS.s.sol:61]: do we have a way to update this after deployment?

File: solidity/script/avs/DeployAVS.s.sol [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+        StrategyParams memory strategyParam;
+
+        uint96 totalMultipliers = 10_000;
+        uint96 multiplier;
+
+        uint96 strategyCount = uint96(strategies.length);
+        for (uint96 i = 0; i < strategyCount; i++) {
+            // the multipliers need to add up to 10,000, so we divide the total by the number of strategies for the first n-1 strategies
+            // and then the last strategy gets the remainder
+            if (i < strategies.length - 1) {
```
Comment (yorhodes) [inline solidity/script/avs/DeployAVS.s.sol:80]: nit: use strategyCount

File: solidity/script/avs/DeployAVS.s.sol [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+                multiplier = totalMultipliers / uint96(strategyCount);
+            } else {
+                multiplier =
+                    totalMultipliers -
+                    multiplier *
+                    uint96(strategyCount - 1);
```
Comment (yorhodes) [inline solidity/script/avs/DeployAVS.s.sol:86]: i dont understand this tbh

File: solidity/script/avs/DeployAVS.s.sol [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+            delegationManager
+        );
+        HyperlaneServiceManager strategyManagerImpl = new HyperlaneServiceManager(
+                address(avsDirectory),
+                address(stakeRegistryImpl),
+                address(paymentCoordinator),
+                address(delegationManager)
+            );
+
+        TransparentUpgradeableProxy hsmProxy = new TransparentUpgradeableProxy(
```
Comment (yorhodes) [inline solidity/script/avs/DeployAVS.s.sol:113]: thanks

File: solidity/script/avs/eigenlayer_addresses.json [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        "strategy": "0x93c4b944D05dfe6df7645A86cd2206016c51564D"
+      },
+      {
+        "name": "Beacon Chain ETH",
+        "strategy": "0xbeaC0eeEeeeeEEeEeEEEEeeEEeEeeeEeeEEBEaC0"
+      }
+    ]
+  },
+  "holesky": {
+    "proxyAdmin": "",
```
Comment (yorhodes) [inline solidity/script/avs/eigenlayer_addresses.json:23]: afaict the current deploy script would set the proxy admin to the empty address?

**Review by yorhodes — APPROVED** [review]
lgtm

File: solidity/script/avs/DeployAVS.s.sol (hunk 24 above)
Comment (aroralanuk) [reply solidity/script/avs/DeployAVS.s.sol:61]: Nope, this is what I was asking Steven for.

File: solidity/script/avs/DeployAVS.s.sol (hunk 26 above)
Comment (aroralanuk) [reply solidity/script/avs/DeployAVS.s.sol:86]: I'll give you an example, 

say strategyCount = 3, then, then the strategy weights are 3333, 3333, 3334 to add upto to 10000.

File: solidity/script/avs/eigenlayer_addresses.json (hunk 28 above)
Comment (aroralanuk) [reply solidity/script/avs/eigenlayer_addresses.json:23]: yup

File: solidity/script/avs/DeployAVS.s.sol (hunk 24 above)
Comment (aroralanuk) [reply solidity/script/avs/DeployAVS.s.sol:56]: added a method to update it in the ECDSAStrategyBase
